src/runner.py

```diff
@@ -1,53 +1,158 @@
+    busy: bool
+    status: str
+    labels: List[RunnerLabel]
+
+
+class LxdInstanceConfigSource(TypedDict):
+    """Represents configuration for source image in LXD instance."""
+
+    type: str
+    mode: str
+    server: str
+    protocol: str
+    alias: str
+
+
+class LxdInstanceConfig(TypedDict):
+    """Represents configuration for LXD instance."""
+
+    name: str
+    type: str
+    source: LxdInstanceConfigSource
+    ephemeral: bool
+    profiles: List[str]
 
 
 class RunnerInfo:
-    def __init__(self, name, local, remote):
+    """RunnerInfo stores the info on a runner.
+
+    Attributes:
+        name (str): Name of the runner.
+        local (pylxd.models.Instance): Local LXD instance information of runner.
+        remote (RemoteRunner): GitHub API information of runner.
+    """
+
+    def __init__(
+        self, name: str, local: Optional[pylxd.models.Instance], remote: Optional[RemoteRunner]
+    ) -> None:
+        """Construct a instance of runner information.
+
+        Args:
+            name (str): Name of the runner.
+            local (pylxd.models.Instance): Local LXD instance information of runner.
+            remote (RemoteRunner): GitHub API information of runner.
+        """
         self.name = name
         self.local = local
         self.remote = remote
 
     @property
-    def is_idle(self):
-        return self.is_online and self.remote.busy is False
+    def is_idle(self) -> bool:
+        """Whether the runner is idle.
+
+        Returns:
+            bool: Whether the runner is idle.
+        """
+        # Redundant None check for mypy type check.
+        return self.remote is not None and self.is_online and self.remote["busy"] is False
 
     @property
-    def is_offline(self):
-        return self.remote and self.remote.status != "online"
+    def is_offline(self) -> bool:
+        """Whether the runner is offline.
+
+        Returns:
+            bool: Whether the runner is offline.
+        """
+        return self.remote is not None and self.remote["status"] != "online"
 
     @property
-    def is_online(self):
-        return self.remote and self.remote.status == "online"
+    def is_online(self) -> bool:
+        """Whether the runner is online.
+
+        Returns:
+            bool: Whether the runner is online.
+        """
+        return self.remote is not None and self.remote["status"] == "online"
 
     @property
-    def is_local(self):
+    def is_local(self) -> bool:
+        """Whether the runner is local.
+
+        Returns:
+            bool: Whether the runner is local.
+        """
         return self.local is not None
 
     @property
-    def virt_type(self):
+    def virt_type(self) -> Optional[str]:
+        """Return the virtualization type of the runner.
+
+        Returns:
+            Optional[str]: Virtualization type of the runner.
+        """
         if self.is_local:
+            # Assert for mypy type check. This check is already done by `is_local`.
```

Comment:
Out of curiosity, what is the `mypy` error if this `assert` is not there?

Reply:
`src/runner.py:154: error: Item "None" of "Optional[Any]" has no attribute "type"  [union-attr]`